File: __Viejo/funciones/ccProveedores.vb
Imports System.Data.SqlClient

Module ccProveedores

    ' ************************************ FUNCIONES DE CUENTAS CORRIENTES DE PROVEEDORES **********************
    Public Function info_ccProveedor(ByVal id_cc As Integer) As ccProveedor
        Dim tmp As New ccProveedor
        Dim sqlstr As String

        Try
            'Crea y abre una nueva conexión
            abrirdb(serversql, basedb, usuariodb, passdb)

            sqlstr = "SELECT * FROM cc_proveedores WHERE id_cc = '" + id_cc.ToString + "'"

            'Propiedades del SqlCommand
            Dim comando As New SqlCommand
            With comando
                .CommandType = CommandType.Text
                .CommandText = sqlstr
                .Connection = CN
            End With

            Dim da As New SqlDataAdapter 'Crear nuevo SqlDataAdapter
            Dim dataset As New DataSet 'Crear nuevo dataset

            da.SelectCommand = comando

            'llenar el dataset
            da.Fill(dataset, "Tabla")
            tmp.id_cc = dataset.Tables("tabla").Rows(0).Item(0).ToString
            tmp.id_proveedor = dataset.Tables("tabla").Rows(0).Item(1).ToString
            tmp.id_moneda = dataset.Tables("tabla").Rows(0).Item(2).ToString
            tmp.nombre = dataset.Tables("tabla").Rows(0).Item(3).ToString
            tmp.saldo = dataset.Tables("tabla").Rows(0).Item(4).ToString
            tmp.activo = dataset.Tables("tabla").Rows(0).Item(5).ToString
            cerrardb()
            Return tmp
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            tmp.nombre = "error"
            cerrardb()
            Return tmp
        End Try
    End Function

    Public Function addCCProveedor(ByVal cc As ccProveedor) As Boolean
        abrirdb(serversql, basedb, usuariodb, passdb)

        Dim mytrans As SqlTransaction
        Dim Comando As New SqlClient.SqlCommand
        Dim sqlstr As String

        mytrans = CN.BeginTransaction()

        Try
            sqlstr = "INSERT INTO cc_proveedores (id_proveedor, id_moneda, nombre, saldo, activo) VALUES ('" + cc.id_proveedor.ToString + "', '" + cc.id_moneda.ToString + "', '" + cc.nombre + "', '" + cc.saldo.ToString + "', '" + cc.activo.ToString + "')"
            Comando = New SqlClient.SqlCommand(sqlstr, CN)

            Comando.Transaction = mytrans
            Comando.ExecuteNonQuery()

            mytrans.Commit()
            cerrardb()
            Return True
        Catch ex As Exception
            MsgBox(ex.Message)
            cerrardb()
            Return False
        End Try
    End Function

    Public Function updateCCProveedor(ByVal cc As ccProveedor, Optional borra As Boolean = False) As Boolean
        abrirdb(serversql, basedb, usuariodb, passdb)

        Dim mytrans As SqlTransaction
        Dim Comando As New SqlClient.SqlCommand
        Dim sqlstr As String

        mytrans = CN.BeginTransaction()

        Try
            If borra = True Then
                sqlstr = "UPDATE cc_proveedores SET activo = '0' WHERE id_cc = '" + cc.id_cc.ToString + "'"
            Else
                sqlstr = "UPDATE cc_proveedores SET id_proveedor = '" + cc.id_proveedor.ToString + "', id_moneda = '" + cc.id_moneda.ToString + "', nombre = '" + cc.nombre + "', saldo = '" + cc.saldo.ToString + "', activo = '" + cc.activo.ToString +
                                               "' WHERE id_cc = '" + cc.id_cc.ToString + "'"
            End If

            Comando = New SqlClient.SqlCommand(sqlstr, CN)

            Comando.Transaction = mytrans
            Comando.ExecuteNonQuery()

            mytrans.Commit()
            cerrardb()
            Return True
        Catch ex As Exception
            MsgBox(ex.Message)
            cerrardb()
            Return False
        End Try
    End Function

    Public Function borrarccProveedor(ByVal cc As ccProveedor) As Boolean
        abrirdb(serversql, basedb, usuariodb, passdb)

        Dim mytrans As SqlTransaction
        Dim Comando As New SqlClient.SqlCommand
        Dim sqlstr As String

        mytrans = CN.BeginTransaction()

        Try
            sqlstr = "DELETE FROM cc_proveedores WHERE id_cc = '" + cc.id_cc.ToString + "'"
            Comando = New SqlClient.SqlCommand(sqlstr, CN)
            Comando.Transaction = mytrans
            Comando.ExecuteNonQuery()

            mytrans.Commit()
            cerrardb()
            Return True
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            cerrardb()
            Return False
        End Try
    End Function


    Public Function consultaTotalCcProveedor(ByVal id_proveedor As Integer, ByVal fecha_desde As Date, ByVal fecha_hasta As Date) As String
        Dim sqlstr As String = ""
        Dim where As String = ""
        Dim da As New SqlDataAdapter 'Crear nuevo SqlDataAdapter
        Dim dt As New DataTable 'Crear nuevo dataset

        Try
            'Crea y abre una nueva conexión
            abrirdb(serversql, basedb, usuariodb, passdb)

            sqlstr = "SELECT DISTINCT SUM(t.total) AS 'Total' " &
                        "FROM transacciones As t " &
                        "INNER JOIN tipos_comprobantes AS tc ON t.id_tipoComprobante = tc.id_tipoComprobante " &
                        "WHERE t.fecha BETWEEN '" + fecha_desde.ToString("MM/dd/yyyy") + "' AND '" + fecha_hasta.ToString("MM/dd/yyyy") + "' " &
                        "AND t.id_proveedor = " + id_proveedor.ToString + " " &
                        "AND tc.contabilizar = 1"


            'Propiedades del SqlCommand
            Dim comando As New SqlCommand
            With comando
                .CommandType = CommandType.Text
                .CommandText = sqlstr
                .Connection = CN
            End With

            da.SelectCommand = comando

            'llenar el dataset
            da.Fill(dt)
            Return dt.Rows(0).Item(0).ToString
        Catch ex As Exception
            MsgBox(ex.Message.ToString)
            Return ""
        Finally
            cerrardb()
        End Try
    End Function
    ' ************************************ FUNCIONES DE CUENTAS CORRIENTES DE PROVEEDORES **********************
End Module

File: funciones/ccProveedores.cs
using System;
using System.Data;
using System.Linq;

namespace Centrex.Funciones
{

    static class ccProveedores
    {

        // ************************************ FUNCIONES DE CUENTAS CORRIENTES DE PROVEEDORES **********************
        public static CcProveedorEntity info_ccProveedor(int  _ccP)
        {
            try
            {
                using (CentrexDbContext context = new CentrexDbContext())
                {
                    if (_ccP != 0 || _ccP != -1)
                    {
                        return context.CcProveedorEntity.FirstOrDefault(c => c.IdProveedor == _ccP);
                    }
                    else                     {
                        return null;
                    }
                }                
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());                
                return null;
            }
        }

        public static bool addCCProveedor(CcProveedorEntity _ccP)
        {
            try
            {
                using (CentrexDbContext context = new CentrexDbContext())
                {
                    var tmp = new CcProveedorEntity()
                    {
                        IdProveedor = _ccP.IdProveedor,
                        IdMoneda = _ccP.IdMoneda,
                        Nombre = _ccP.Nombre,
                        Saldo = _ccP.Saldo,
                        Activo = _ccP.Activo
                    };

                    context.CcProveedorEntity.Add(tmp);
                    context.SaveChanges();
                    return true;
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message);
                return false;
            }
        }

        public static bool updateCCProveedor(CcProveedorEntity _ccP, bool borra = false)
        {
            try
            {
                using (CentrexDbContext context = new CentrexDbContext())
                {
                    if (_ccP is not null)
                    {
                        if (borra == true)
                        {
                            _ccP.Activo = false;
                        }
                        else
                        {
                            _ccP.IdProveedor = _ccP.IdProveedor;
                            _ccP.IdMoneda = _ccP.IdMoneda;
                            _ccP.Nombre = _ccP.Nombre;
                            _ccP.Saldo = _ccP.Saldo;
                            _ccP.Activo = _ccP.Activo;
                        }

                        context.SaveChanges();
                        return true;
                    }
                    else
                    {
                        return false;
                    }
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message);
                return false;
            }
        }

        public static bool borrarccProveedor(CcProveedorEntity _ccP)
        {
            try
            {
                using (CentrexDbContext context = new CentrexDbContext())
                {
                    if (_ccP is not null)
                    {
                        context.CcProveedorEntity.Remove(_ccP);
                        context.SaveChanges();
                        return true;
                    }
                    else
                    {
                        return false;
                    }
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return false;
            }
        }


        public static string consultaTotalCcProveedor(int id_proveedor, DateOnly fecha_desde, DateOnly fecha_hasta)
        {
            try
            {
                using (CentrexDbContext context = new CentrexDbContext())
                {
                    var importeTotal = (from trans in context.TransaccionEntity
                                        join tipoComp in context.TipoComprobanteEntity on trans.IdTipoComprobante equals tipoComp.IdTipoComprobante
                                        where trans.Fecha >= fecha_desde && trans.Fecha <= fecha_hasta && trans.IdProveedor == id_proveedor && tipoComp.Contabilizar == true
                                        select trans.Total).Distinct().Sum();

                    // Manejo de valores nulos
                    if (importeTotal is null)
                    {
                        return "0";
                    }
                    else
                    {
                        return importeTotal.ToString();
                    }
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox(ex.Message.ToString());
                return "";
            }
        }

        // ************************************ FUNCIONES DE CUENTAS CORRIENTES DE PROVEEDORES **********************
    }
}
